text/0000-really-tagged-unions.md

```diff
@@ -0,0 +1,362 @@
+
+In addition, it is defined for Rust programs to cast/reinterpret/transmute such an enum into the equivalent tag+union Rust definition above. Seperately manipulating the tag and payload is also defined. The tag and payload need only be in a consistent/initialized state when the value is matched on (which includes Dropping it). This means that the contents of a `#[repr(X)]` enum cannot be inspected by outer enums for storage optimizations -- `Option<MyEnum>` must be larger than `MyEnum`.
+
+For instance, this code is valid (using the same definitions above):
+
+```rust
+/// Tries to parse a MyEnum from a custom binary format, overwriting `dest`.
+/// On Err, `dest` may be partially overwritten (but will be in a memory-safe state)
+fn parse_my_enum_from<'a>(dest: &'a mut MyEnum, input: &mut &[u8]) -> Result<(), &'static str> {
+    unsafe {
+        // Convert to raw repr
+        let dest: &'a mut MyEnumRepr = mem::transmute(dest);
+
+        // If MyEnum was non-trivial, we might match on the tag and 
+        // drop_in_place the payload here to start.
+
+        // Read the tag
+        let tag = input.get(0).ok_or("Couldn't Read Tag")?;
+        dest.tag = match tag {
+            0 => MyEnumTag::A,
+            1 => MyEnumTag::B,
+            2 => MyEnumTag::C,
+            3 => MyEnumTag::D,
+            _ => { return Err("Invalid Tag Value"); }
+        };
+        *input = &input[1..];
+
+        // Note: it would be very bad if we panicked past this point, or if
+        // the following methods didn't initialize the payload on Err!
+
+        // Read the payload
+        match dest.tag {
+            MyEnumTag::A => parse_my_enum_a_from(&mut dest.payload.A, input),
+            MyEnumTag::B => parse_my_enum_b_from(&mut dest.payload.B, input),
+            MyEnumTag::C => parse_my_enum_c_from(&mut dest.payload.C, input),
+            MyEnumTag::D => { Ok(()) /* do nothing */ }
+        }
+    }
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Since the whole point of this proposal is to enable low-level control, the guide-level explanation should cover all the relevant corner-cases and details in sufficient detail. All that remains is to discuss implementation details. Thankfully, this is very simple: unless there's been an error, this should exactly match the current behaviour of the compiler.
+
+It was [informally decided earilier this year](https://github.com/rust-lang/rust/issues/40029) that we should have this behaviour, as it was being relied on and it made sense. There is even a test in the rust-lang repo that was added to ensure that this behaviour doesn't regress.
+
+We may need to double-check that the sub-structs are actually being marked as `repr(C)`.
+
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There aren't really any drawbacks: we already decided to do this. This is just making it official.
```

Comment:
Saying "it has been decided already" is neither a good attitude towards the RFC process, nor is it a substitute for a drawback section!